README.txt
bigbluebutton
-------------

The work of this project is derivated from https://github.com/schallis/django-bigbluebutton 98f2259fa3 by Steve Challis.

It is a wrapper for accessing the API of bigbluebutton http://code.google.com/p/bigbluebutton/wiki/API


A simple example ::

import argparse
from bigbluebutton import Meeting_Setup, Meeting
import bbb_settings 


if __name__ == '__main__':
    PARSER = argparse.ArgumentParser(description='creates and join a session')
    PARSER.add_argument('--meeting_name', dest="meeting_name", type=str, required=True,
                       help='name of the meeting')
    PARSER.add_argument('--meeting_id', dest='meeting_id', required=True,
                       help='id for the meeting')
    PARSER.add_argument('--moderator', dest='moderator', required=True,
                       help='name of the meeting moderator')
    PARSER.add_argument( '--moderator_password', dest='moderator_password', required=True,
                       help='password for moderator')
    PARSER.add_argument( '--attendee_password', dest='attendee_password', required=True,
                       help='password for attendee')
    PARSER.add_argument( '--url', dest='url', required=True,
                       help='pre upload url')

    ARGS = PARSER.parse_args()

    session = Meeting_Setup(bbb_settings.BBB_API_URL, bbb_settings.SALT,
                            ARGS.meeting_name, ARGS.meeting_id,
                            ARGS.attendee_password, ARGS.moderator_password,
                            pre_upload_slide=ARGS.url)
    session.create_meeting()
    print "meeting expires if noone joins in"
    
    meeting = Meeting(bbb_settings.BBB_API_URL, bbb_settings.SALT)
    print "MODERATOR:"
    print meeting.join_url(ARGS.meeting_id, ARGS.moderator, ARGS.moderator_password)
    print '-------------------------------------------'

    print "RANDOM USER:"
    print meeting.join_url(ARGS.meeting_id, 'RANDOM', ARGS.attendee_password)
    print '-------------------------------------------'    

    print "ALL MEETINGS"
    print meeting.get_meetings()
    print '-------------------------------------------'

    print "IS RUNNING (meeting is only running after someone has joined in)"
    print meeting.is_running(ARGS.meeting_id)
    print '-------------------------------------------'

    print "END MEETING URL"
    print meeting.end_meeting_url(ARGS.meeting_id, ARGS.moderator_password)
    print '-------------------------------------------'


    if meeting.is_running(ARGS.meeting_id):
        print "END MEETING"
        meeting.end_meeting(ARGS.meeting_id, ARGS.moderator_password)
        print '-------------------------------------------'
    


CHANGES.txt
0.5.0
--------
PEP8 fixes
Meeting_Setup renamed to MeetingSetup
reduced code duplication


0.4.1
-------
small fixes in Type Comparison

0.4.0
-----
urllib2 replaced by requests
pre upload of slides by URL added
further attribute for recording


0.3.0
------
extended and refactored meeting_info to get all 0.8 attributes
API calls for recordings added


0.2.1
-----

Meeting_Setup: logout_url parameter changed from None to ''.
               This points to the default url given in the config of bigbluebutton.


0.2.0
-----
Meeting_Setup: logout_url, max_participants, duration and welcome parameter added
Meeting: end_meeting_url added

0.1.1
-----
minor bug fix

0.1.0
-----
initial version of bigbluebutton python api with some creation parameters
based on https://github.com/schallis/django-bigbluebutton 98f2259fa3 by Steve Challis.



bigbluebutton/utils.py
# Copyright: 2011 Steve Challis (http://schallis.com)
# Copyright: 2012-2016 MoinMoin:ReimarBauer
# License: MIT 

"""
    bigbluebutton.utils

    This module contains helper functions to access bigbluebutton servers

"""
import requests
import xml.etree.ElementTree as ET
from hashlib import sha1

def parse(response):
    """
    :param reponse: XML Data 
    """
    try:
        xml = ET.fromstring(response)
        code = xml.find('returncode').text
        if code == 'SUCCESS':
            return xml
        else:
            raise
    except ET.ParseError:
        return None

def api_call(salt, query, call):
    """
    builds the hash based on the call, query and salt
    
    :param salt: The security salt defined for your bigbluebutton instance
    :param query: The query parameters for calling the bigbluebutton resource
    :param call: The bigbluebutton resource name
    """
    prepared = "%s%s%s" % (call, query, salt)
    checksum = sha1(prepared).hexdigest()
    return "%s&checksum=%s" % (query, checksum)

def get_xml(bbb_api_url, salt, call, query, pre_upload_slide=None):
    """
    gets XML from the bigbluebutton ressource

    :param bbb_api_url: The url to your bigbluebutton instance (including the api/)
    :param salt: The security salt defined for your bigbluebutton instance
    :param call: The bigbluebutton resource name
    :param query: The query parameters for calling the bigbluebutton resource
    :param pre_upload_slide: on create a file could be uploaded
    """

    hashed = api_call(salt, query, call)
    url = bbb_api_url + call + '?' + hashed
    if call == "create" and pre_upload_slide is not None:
        xml = "<?xml version='1.0' encoding='UTF-8'?> <modules> <module name='presentation'> <document url='%(pre_upload_slide)s'/> </module></modules>" % {"pre_upload_slide": pre_upload_slide}
        headers = {'Content-Type': 'application/xml'}
        return parse(requests.post(url, data=xml, headers=headers).content)
    else:
        return parse(requests.get(url).content)


def xml_match(xml, match):
    """
    Finds the first subelement matching match and verifies that its text attribute is 'true'

    :param xml: xml as ET instance
    :param match: pattern to lookup by find
    :return: boolean
    """
    if xml is not None:
        return xml.find(match).text == 'true'
    return False


bigbluebutton/__init__.py
# Copyright: 2011 Steve Challis (http://schallis.com)
# Copyright: 2012-2016 MoinMoin:ReimarBauer
# License: MIT

"""
    bigbluebutton



    This module contains functions to access bigbluebutton servers
    Meeting_Setup: for initializing a meeting.
    Meeting: for operations on the meeting after initializing.

"""
import random
from urllib import urlencode
from bigbluebutton.utils import api_call, get_xml, xml_match


class Meeting_Setup(object):
    """
    Initializes meetings
    """

    def __init__(self, bbb_api_url=None, salt=None, meeting_name='', meeting_id='',
                 attendee_password=None, moderator_password=None,
                 logout_url='', max_participants=-1, duration=0, dial_number='',
                 welcome=u'Welcome!',
                 moderator_only_message=u'', meta=u'',
                 record=False, auto_start_recording=False, allow_start_stop_recording=True,
                 pre_upload_slide=None
                 ):
        """
        :param bbb_api_url: The url to your bigbluebutton instance (including the api/)
        :param salt: The security salt defined for your bigbluebutton instance
        :param meeting_name: A name for the meeting.
        :param meeting_id: A meeting ID that can be used to identify this meeting by the third party application.
                           This must be unique to the server that you are calling. If you supply a non-unique
                           meeting ID, you will still have a successful call, but will receive a warning message
                           in the response.
                           If you intend to use the recording feature, the meetingID shouldn't contain commas.
        :param attendee_password: The password that will be required for attendees to join the meeting.
                                  This is optional, and if not supplied, BBB will assign a random password.
        :param moderator_password:  The password that will be required for moderators to join the meeting or
                                    for certain administrative actions (i.e. ending a meeting). This is optional,
                                    and if not supplied, BBB will assign a random password.
        :param logout_url: The URL that the BigBlueButton client will go to after users click the OK button on
                           the 'You have been logged out message'. This overrides, the value for
                           bigbluebutton.web.loggedOutURLif defined in bigbluebutton.properties
        :param max_participants: The maximum number of participants to allow into the meeting (including moderators).
                                 After this number of participants have joined, BBB will return an appropriate error
                                 for other users trying to join the meeting. A negative number indicates that
                                 an unlimited number of participants should be allowed (this is the default setting).
        :param duration: The duration parameter allows to specify the number of minutes for the meeting's length.
                         When the length of the meeting reaches the duration, BigBlueButton automatically
                         ends the meeting. The default is 0, which means the meeting continues until the last person
                         leaves or an end API calls is made with the associated meetingID.
        :param dial_number: The dial access number that participants can call in using regular phone.
        :param welcome: A welcome message that gets displayed on the chat window when the participant joins.
                        You can include keywords (%%CONFNAME%%, %%DIALNUM%%, %%CONFNUM%%) which will be
                        substituted automatically. You can set a default welcome message on bigbluebutton.properties
        :param moderator_only_message: Display a message to all moderators in the public chat.
        :param meta: You can pass one or more metadata values for create a meeting.
                    These will be stored by BigBlueButton and later retrievable via the getMeetingInfo call
                    and getRecordings. Examples of meta parameters are meta_Presenter, meta_category, meta_LABEL, etc.
                    All parameters are converted to lower case, so meta_Presenter would be the same as meta_PRESENTER.
        :param record: Setting record=True instructs the BigBlueButton server to record the media and events in
                       the session for later playback. Available values are true or false. Default value is false.
        :param auto_start_recording: Default=False, Setting start_recording=True will automatically
                                     starts recording when first user joins.
                                     NOTE: Don't set to autoStartRecording =false and allowStartStopRecording=false
                                     as the user won't be able to record.
        :param allow_start_stop_recording: Default=True, Allow the user to start/stop recording.
                                           This means the meeting can start recording automatically
                                           (autoStartRecording=true) with the user able to stop/start
                                           recording from the client.
        :param pre_upload_slide: You can preupload slides within the create call by providing an URL to the slides.
        """
        self.bbb_api_url = bbb_api_url
        self.salt = salt
        self.meeting_name = meeting_name
        self.meeting_id = meeting_id
        self.attendee_password = attendee_password
        self.moderator_password = moderator_password
        self.logout_url = logout_url
        self.max_participants = max_participants
        self.duration = duration
        self.dial_number = dial_number
        self.welcome = welcome
        self.record = str(record).lower()
        self.moderator_only_message = moderator_only_message
        self.meta = meta
        self.allow_start_stop_recording = allow_start_stop_recording
        self.auto_start_recording = auto_start_recording
        self.pre_upload_slide = pre_upload_slide

    def create_meeting(self):
        """
        creates the meeting
        """
        if not Meeting(self.bbb_api_url, self.salt).is_running(self.meeting_id):
            call = 'create'
            voicebridge = 70000 + random.randint(0, 9999)
            query = urlencode((
                ('name', self.meeting_name),
                ('meetingID', self.meeting_id),
                ('attendeePW', self.attendee_password),
                ('moderatorPW', self.moderator_password),
                ('voiceBridge', voicebridge),
                ('dialNumber', self.dial_number),
                ('welcome', self.welcome),
                ('logoutURL', self.logout_url),
                ('maxParticipants', self.max_participants),
                ('duration', self.duration),
                ('record', self.record),
                ('meta', self.meta),
                ('moderatorOnlyMessage', self.moderator_only_message),
                ('autoStartRecording', self.auto_start_recording),
                ('allowStartStopRecording', self.allow_start_stop_recording),
            ))
            xml = get_xml(self.bbb_api_url, self.salt, call, query, self.pre_upload_slide)
            return xml is not None


class Meeting(object):
    """
    gives access to meetings
    """

    def __init__(self, bbb_api_url=None, salt=None):
        """
        :param bbb_api_url: The url to your bigbluebutton instance (including the api/)
        :param salt: The security salt defined for your bigbluebutton instance
        """
        self.bbb_api_url = bbb_api_url
        self.salt = salt

    def is_running(self, meeting_id):
        """
        This call enables you to simply check on whether or not a meeting is
        running by looking it up with your meeting ID.

        :param meeting_id: ID that can be used to identify the meeting
        """
        call = 'isMeetingRunning'
        match = 'running'
        query = urlencode((
            ('meetingID', meeting_id),
        ))
        xml = get_xml(self.bbb_api_url, self.salt, call, query)
        return xml_match(xml, match)

    def join_url(self, meeting_id, name, password):
        """
        generates the url for accessing a meeting

        :param meeting_id: ID that can be used to identify the meeting
        :param name: The name that is to be used to identify this user to
                     other conference attendees.
        :param password: The password that this attendee is using.
                         If the moderator password is supplied, he will be
                         given moderator status
                         (and the same for attendee password, etc)
        """
        call = 'join'
        query = urlencode((
            ('fullName', name),
            ('meetingID', meeting_id),
            ('password', password),
        ))
        hashed = api_call(self.salt, query, call)
        return self.bbb_api_url + call + '?' + hashed

    def end_meeting_url(self, meeting_id, password):
        """
        Use this to generate the url to end a meeting

        :param meetingID: The meeting ID that identifies the meeting you are attempting to end.
        :param password: The moderator password for this meeting.
                         You can not end a meeting using the attendee password.
        """
        call = 'end'
        query = urlencode((
            ('meetingID', meeting_id),
            ('password', password),
        ))

        hashed = api_call(self.salt, query, call)
        return self.bbb_api_url + call + '?' + hashed

    def end_meeting(self, meeting_id, password):
        """
        Use this to forcibly end a meeting and kick all participants out of the meeting.

        :param meetingID: The meeting ID that identifies the meeting you are attempting to end.
        :param password: The moderator password for this meeting.
                         You can not end a meeting using the attendee password.
        """
        call = 'end'
        query = urlencode((
            ('meetingID', meeting_id),
            ('password', password),
        ))
        xml = get_xml(self.bbb_api_url, self.salt, call, query)
        return xml is not None

    def meeting_info(self, meeting_id, password):
        """
        This call will return all of a meeting's information,
        including the list of attendees as well as start and end times.

        :param meetingID: The meeting ID that identifies the meeting
        :param password: The moderator password for this meeting.
        """
        call = 'getMeetingInfo'
        query = urlencode((
            ('meetingID', meeting_id),
            ('password', password),
        ))
        xml = get_xml(self.bbb_api_url, self.salt, call, query)
        if xml is not None:
            # Create dict of values for easy use in template
            users = []
            attendees = xml.find('attendees')
            if attendees is not None:
                for attendee in attendees.getchildren():
                    user = {}
                    user['user_id'] = attendee.find('userID').text
                    user['name'] = attendee.find('fullName').text
                    user['role'] = attendee.find('role').text
                    users.append(user)

            meeting_info = {
                'meeting_name': xml.find('meetingName').text,
                'meeting_id': xml.find('meetingID').text,
                'create_time': int(xml.find('createTime').text),
                'voice_bridge': int(xml.find('voiceBridge').text),
                'attendee_pw': xml.find('attendeePW').text,
                'moderator_pw': xml.find('moderatorPW').text,
                'running': xml.find('running').text == "true",
                'recording': xml.find('recording').text == "true",
                'has_been_forcibly_ended': xml.find('hasBeenForciblyEnded').text == "true",
                'start_time': int(xml.find('startTime').text),
                'end_time': int(xml.find('endTime').text),
                'participant_count': int(xml.find('participantCount').text),
                'max_users': int(xml.find('maxUsers').text),
                'moderator_count': int(xml.find('moderatorCount').text),
                'users': users
            }
            return meeting_info
        else:
            return None

    def get_meetings(self):
        """
        This call will return a list of all the meetings found on this server.
        """
        call = 'getMeetings'
        query = urlencode((
            ('random', 'random'),
        ))

        xml = get_xml(self.bbb_api_url, self.salt, call, query)
        if xml is not None:
            # Create dict of values for easy use in template
            all_meetings = []
            meetings = xml[1].findall('meeting')
            for meeting in meetings:
                meeting_id = meeting.find('meetingID').text
                password = meeting.find('moderatorPW').text
                all_meetings.append({
                    'name': meeting_id,
                    'moderator_pw': password,
                    'attendee_pw': meeting.find('attendeePW').text,
                    'has_been_forcibly_ended': meeting.find('hasBeenForciblyEnded').text == "true",
                    'running': meeting.find('running').text == "true",
                    'create_time': int(meeting.find('createTime').text),
                    'info': self.meeting_info(
                        meeting_id,
                        password)
                })
            return all_meetings
        else:
            return None

    def get_recordings(self, meeting_id):
        """
        Retrieves the recordings that are available for playback for a given meetingID (or set of meeting IDs).

        :param meetingID: The meeting ID that identifies the meeting
        """
        call = 'getRecordings'
        query = urlencode((
            ('meetingID', meeting_id),
        ))
        xml = get_xml(self.bbb_api_url, self.salt, call, query)
        # ToDO implement more keys
        if xml is not None:
            recordings = xml.find('recording')
            records = []
            for meeting in recordings.getchildren():
                record = {}
                record['record_id'] = meeting.find('recordID').text
                record['meeting_id'] = meeting.find('meetingID').text
                record['meeting_name'] = meeting.find('name').text
                record['published'] = meeting.find('published').text == "true"
                record['start_time'] = meeting.find('startTime').text
                record['end_time'] = meeting.find('endTime').text
                records.append(record)
            return records
        else:
            return None

    def publish_recordings(self, record_id, publish=False):
        """
        Publish and unpublish recordings for a given recordID (or set of record IDs).

        :param record_id: A record ID for specify the recordings to apply the publish action.
                         It can be a set of meetingIDs separate by commas.
        :param publish: The value for publish or unpublish the recording(s). Available values: True or False.
        """
        call = 'publishRecordings'
        match = 'published'
        query = urlencode((
            ('recordID', record_id),
            'publish', str(publish).lower()
        ))
        xml = get_xml(self.bbb_api_url, self.salt, call, query)
        return xml_match(xml, match)

    def delete_recordings(self, record_id, publish=False):
        """
        Delete one or more recordings for a given recordID (or set of record IDs).

        :param record_id: A record ID for specify the recordings to delete. It can be a set of
                         meetingIDs separate by commas.
        :param publish: The value for publish or unpublish the recording(s). Available values: True or False.
        """
        call = 'deleteRecordings'
        match = "deleted"
        query = urlencode((
            ('recordID', record_id),
            'publish', str(publish).lower()
        ))
        xml = get_xml(self.bbb_api_url, self.salt, call, query)
        return xml_match(xml, match)


BBB.py

import argparse
from urllib2 import urlopen
from urllib import urlencode
from hashlib import sha1
import xml.etree.ElementTree as ET
import random
import bbb_settings as settings

def parse(response):
    try:
        xml = ET.XML(response)
        code = xml.find('returncode').text
        if code == 'SUCCESS':
            return xml
        else:
            raise
    except:
        return None

def api_call(query, call):
    prepared = "%s%s%s" % (call, query, settings.SALT)
    checksum = sha1(prepared).hexdigest()
    result = "%s&checksum=%s" % (query, checksum)
    return result

def join_url(meeting_id, name, password):
    call = 'join'
    query = urlencode((
                       ('fullName', name),
                       ('meetingID', meeting_id),
                       ('password', password),
                       ))
    hashed = api_call(query, call)
    url = settings.BBB_API_URL + call + '?' + hashed
    return url


class Meeting(object):
    def __init__(self, meeting_name='', meeting_id='', attendee_password=None, moderator_password=None):
        self.meeting_name = meeting_name
        self.meeting_id = meeting_id
        self.attendee_password = attendee_password
        self.moderator_password = moderator_password

    def is_running(self):
        call = 'isMeetingRunning'
        query = urlencode((
            ('meetingID', self.meeting_id),
        ))
        hashed = api_call(query, call)
        url = settings.BBB_API_URL + call + '?' + hashed
        result = parse(urlopen(url).read())
        if result:
            return result.find('running').text
        else:
            return 'error'

    @classmethod
    def end_meeting(self, meeting_id, password):
        call = 'end'
        query = urlencode((
            ('meetingID', meeting_id),
            ('password', password),
        ))
        hashed = api_call(query, call)
        url = settings.BBB_API_URL + call + '?' + hashed
        result = parse(urlopen(url).read())
        if result:
            pass
        else:
            return 'error'

    @classmethod
    def meeting_info(self, meeting_id, password):
        call = 'getMeetingInfo'
        query = urlencode((
            ('meetingID', meeting_id),
            ('password', password),
        ))
        hashed = api_call(query, call)
        url = settings.BBB_API_URL + call + '?' + hashed
        r = parse(urlopen(url).read())
        if r:
            # Create dict of values for easy use in template
            d = {
                'start_time': r.find('startTime').text,
                'end_time': r.find('endTime').text,
                'participant_count': r.find('participantCount').text,
                'moderator_count': r.find('moderatorCount').text,
                'moderator_pw': r.find('moderatorPW').text,
                'attendee_pw': r.find('attendeePW').text,
                'invite_url': 'join=%s' % meeting_id,
            }
            return d
        else:
            return None

    @classmethod
    def get_meetings(self):
        call = 'getMeetings'
        query = urlencode((
            ('random', 'random'),
        ))
        hashed = api_call(query, call)
        url = settings.BBB_API_URL + call + '?' + hashed
        result = parse(urlopen(url).read())
        if result:
            # Create dict of values for easy use in template
            d = []
            r = result[1].findall('meeting')
            for m in r:
                meeting_id = m.find('meetingID').text
                password = m.find('moderatorPW').text
                d.append({
                    'name': meeting_id,
                    'running': m.find('running').text,
                    'moderator_pw': password,
                    'attendee_pw': m.find('attendeePW').text,
                    'info': Meeting.meeting_info(
                        meeting_id,
                        password)
                })
            return d
        else:
            return 'error'

    def create_meeting(self):
        call = 'create' 
        voicebridge = 70000 + random.randint(0,9999)
        query = urlencode((
            ('name', self.meeting_name),
            ('meetingID', self.meeting_id),
            ('attendeePW', self.attendee_password),
            ('moderatorPW', self.moderator_password),
            ('voiceBridge', voicebridge),
            ('welcome', "Welcome!"),
        ))
        hashed = api_call(query, call)
        url = settings.BBB_API_URL + call + '?' + hashed
        result = parse(urlopen(url).read())
        if result:
            return result
        else:
            raise


if __name__ == '__main__':
    PARSER = argparse.ArgumentParser(description='creates and join a session')
    PARSER.add_argument('--meeting_name', dest="meeting_name", type=str, required=True,
                       help='name of the meeting')
    PARSER.add_argument('--meeting_id', dest='meeting_id', required=True,
                       help='id for the meeting')
    PARSER.add_argument('--moderator', dest='moderator', required=True,
                       help='name of the meeting moderator')
    PARSER.add_argument( '--moderator_password', dest='moderator_password', required=True,
                       help='password for moderator')
    PARSER.add_argument( '--attendee_password', dest='attendee_password', required=True,
                       help='password for attendee')

    ARGS = PARSER.parse_args()

    SESSION = Meeting(ARGS.meeting_name, ARGS.meeting_id, ARGS.attendee_password, ARGS.moderator_password)
    SESSION.create_meeting()
    print "MODERATOR:"
    print join_url(ARGS.meeting_id, ARGS.moderator, ARGS.moderator_password)
    print '-------------------------------------------'

    print "RANDOM USER:"
    print join_url(ARGS.meeting_id, 'RANDOM', ARGS.attendee_password)
    print '-------------------------------------------'    

    print "ALL MEETINGS"
    print SESSION.get_meetings()
    print '-------------------------------------------'

    print "IS RUNNING (meeting is only running after someone has joined in)"
    print SESSION.is_running()
    print '-------------------------------------------'

    print "END MEETING"
    SESSION.end_meeting(ARGS.meeting_id, ARGS.moderator_password)
    print '-------------------------------------------'
    
    
    
